src/cpp/r/RSexp.cpp

```diff
@@ -838,6 +896,164 @@ bool inherits(SEXP object, const char* S3Class)
+   nsePrimitives.insert("library");
+   nsePrimitives.insert("require");
+   nsePrimitives.insert("enquote");
+   nsePrimitives.insert("bquote");
+   nsePrimitives.insert("expression");
+   nsePrimitives.insert("evalq");
+   nsePrimitives.insert("subset");
+   return nsePrimitives;
+}
+
+const std::set<std::string>& nsePrimitives()
+{
+   static const std::set<std::string> set = makeNsePrimitives();
+   return set;
+}
+
+// Attempts to find calls to functions which perform NSE.
+bool maybePerformsNSEImpl(SEXP node,
+                          const std::set<std::string>& nsePrimitives)
+{
+   if (TYPEOF(node) == LANGSXP)
+   {
+      // Check to see if this is a call to a function typically
+      // used with non-standard evaluation
+      SEXP head = CAR(node);
+      if (TYPEOF(head) == SYMSXP && nsePrimitives.count(CHAR(PRINTNAME(head))))
+         return true;
+      
+      // Iterate over all other elements in the pairlist and check to see if
+      // they make calls to NSE primitives.
+      SEXP tail = CDR(node);
+      head = CAR(tail);
+      while (tail != R_NilValue)
+      {
+         if (TYPEOF(head) == LANGSXP && maybePerformsNSEImpl(head, nsePrimitives))
+            return true;
+         
+         tail = CDR(tail);
+         head = CAR(tail);
+      }
+   }
+   
+   return false;
+}
+
+bool maybePerformsNSE(SEXP function)
+{
+   if (!Rf_isFunction(function))
+      return false;
+   
+   if (Rf_isPrimitive(function))
+      return false;
+   
+   return maybePerformsNSEImpl(
+            BODY_EXPR(function),
+            nsePrimitives());
+}
+
+// NOTE: Uses `R_lsInternal` which throws error if a non-environment is
+// passed; we therefore perform this validation ourselves before calling
+// `R_lsInternal`.
+SEXP objects(SEXP environment,
+             bool allNames,
+             Protect* pProtect)
+{
+   if (TYPEOF(environment) != ENVSXP)
+   {
+      LOG_ERROR_MESSAGE("'objects' called on non-environment");
+      return R_NilValue;
+   }
+   
+   SEXP resultSEXP;
+   pProtect->add(resultSEXP = R_lsInternal(environment, allNames ? TRUE : FALSE));
+   return resultSEXP;
+}
+
+Error objects(SEXP environment,
+              bool allNames,
+              std::vector<std::string>* pNames)
+{
+   Protect protect;
+   SEXP objectsSEXP = objects(environment, allNames, &protect);
+   
+   if (Rf_isNull(objectsSEXP))
+      return Error(errc::CodeExecutionError, ERROR_LOCATION);
+   
+   if (!fillVectorString(objectsSEXP, pNames))
+      return Error(errc::CodeExecutionError, ERROR_LOCATION);
+   
+   return Success();
+}
+
+SEXP mget(SEXP objectNames,
+          SEXP environment,
+          bool inherits,
+          Protect* pProtect)
+{
+   SEXP resultSEXP;
+   
+   r::exec::RFunction mget("mget");
+   mget.addParam("x", objectNames);
+   mget.addParam("envir", environment);
+   mget.addParam("inherits", inherits);
+   
+   Error error = mget.call(&resultSEXP, pProtect);
+   IF_ERROR(error, return R_NilValue);
```

Comment:
I really dislike the IF_ERROR business. It makes me have to pause and think every time I scan over code that includes it. Here's another way to do this from our code:

https://github.com/rstudio/rstudio/blob/41b8cc23789d776a2ef5668c00b77fda941c7abb/src/cpp/core/system/System.cpp#L67

If I have an error in a `main()` function I can just call:

return exitFailure(error, ERROR_LOCATION);

Note that you'd need a version of this for every return type but my guess is that there aren't that many (false, R_NilValue, error itself, etc.)